pkg/core/klinebuilder.go

```diff
@@ -0,0 +1,221 @@
+			kb.minInterval,
+			interval,
+		)
+		kb.logger.Error(err)
+		panic(err)
+	}
+	kb.intervals[interval] = struct{}{}
+}
+
+func (kb *kLineBuilder) BindStream(stream types.Stream) {
+	if kb.isBackTesting {
+		stream.OnKLineClosed(kb.EmitKLineClosed)
+		stream.OnKLine(kb.EmitKLine)
+		return
+	}
+	stream.OnMarketTrade(kb.handleMarketTrade)
+}
+
+func (kb *kLineBuilder) SetExchange(exchange types.Exchange) {
+	kb.exchange = exchange
+}
+
+func (kb *kLineBuilder) Run(ctx context.Context) {
+	if kb.running {
+		panic("can not run kline builder twice")
+	}
+	if len(kb.intervals) == 0 || kb.isBackTesting {
+		return
+	}
+	kb.logger.Infof("kline updater started for %s (%+v)", kb.symbol, kb.intervals)
+	kb.running = true
+	go kb.run(ctx)
+}
+
+func (kb *kLineBuilder) handleMarketTrade(trade types.Trade) {
+	// the market data stream may subscribe to trades of multiple symbols
+	// so we need to check if the trade is for the symbol we are interested in
+	if trade.Symbol != kb.symbol {
+		return
+	}
+	kb.mu.Lock()
+	defer kb.mu.Unlock()
+
+	kb.state.AddTrade(trade)
+}
+
+func (kb *kLineBuilder) run(ctx context.Context) {
+	startTime := time.Now()
+	startTimeTruncated := startTime.Truncate(kb.minInterval.Duration())
+	intervalDuration := kb.minInterval.Duration()
+	if !startTimeTruncated.Equal(startTime) {
+		// wait for the next interval
+		// it's necessary since we want to collect trades as accurately as possible
+		waitDuration := intervalDuration - startTime.Sub(startTimeTruncated)
+		select {
+		case <-time.After(waitDuration):
+		case <-ctx.Done():
+			return
+		}
+	}
+	// start ticker
+	ticker := time.NewTicker(intervalDuration)
+	defer ticker.Stop()
+
+	// reset the state
+	// all kline of the intervals will be reset to the start time
+	kb.mu.Lock()
+	startTime = time.Now().Truncate(intervalDuration)
+	for interval := range kb.intervals {
+		kb.state.Reset(interval, types.Time(startTime))
+	}
+	kb.mu.Unlock()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case tickTime := <-ticker.C:
+			kb.tick(tickTime)
+		}
+	}
+}
+
+func (kb *kLineBuilder) tick(tickTime time.Time) {
+	kb.mu.Lock()
+	defer kb.mu.Unlock()
+
+	for interval := range kb.intervals {
+		okForEmit := false
+		kline, exist := kb.state.GetKLine(interval)
+		if !exist {
+			// should not happen
+			continue
+		}
+		if kline.NumberOfTrades == 0 {
+			// no trades in this interval
+			// check if we have a last kline
+			lastKline, found := kb.state.GetLastKLine(interval)
+			oriStarTime := kline.StartTime
+			if found {
+				kline.Set(lastKline)
+				kline.StartTime = oriStarTime
+				okForEmit = true
+			} else if kb.exchange != nil {
+				// try to query kline from RESTful API
+				queryStartTime := kline.StartTime.Time().Add(-kline.Interval.Duration())
+				queryEndTime := queryStartTime.Add(2 * kline.Interval.Duration())
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				klines, err := kb.exchange.QueryKLines(
```

Comment:
we can't call QueryKlines in the back-test mode, and it seems like you're mixing 2 difference scenario in the same component?